### PR TITLE
SI-7214 outer check based on dealiased pattern type.

### PR DESCRIPTION
**Pull request by retronym**
A Typed Pattern (`_: T`) is more than `.isInstanceOf`: if `T` is a
path dependent type, the scrutinee's $outer reference is also
compared against the prefix of `T`.

The code that synthesises this is split into two places.
`needsOuterCheck` determines whether to add this check, based on
the type `T`, and the type of the scrutinee. If it gives the
go-ahead, `treeCondStrategy.outerCheck` synthesizes the check.

The new test case demonstrates the problems caused by the failure
to dealias in `needsOuterCheck`: it could either wrongly lead to
synthesis of an outer test (which would crash), or wrongly omit
the outer test (meaning overly liberal matching.)

A simple `dealias` remedies this. `dealiasWiden` is _not_
appropriate here; we need to keep hold of singleton types.

I'll also note that there is already a little slack between these
methods, as commented:

> ExplicitOuter replaces `Select(q, outerSym) OBJ_EQ expectedPrefix`
> by `Select(q, > outerAccessor(outerSym.owner)) OBJ_EQ expectedPrefix`
> if there's an outer accessor, otherwise the condition becomes `true`
> TODO: can we improve needsOuterTest so there's always an outerAccessor?

So this is probably a fragile area that warrants a careful review
with a view to design improvements.

Review by @adriaanm

### CODE REVIEW

**Comment by paulp** [comment]
Also Si-6813, right? "dealias types when looking for uncheckable outer checks" ?

**Comment by paulp** [comment]
LGTM

File: test/files/run/t7214.scala [hunk 1]
```diff
@@ -0,0 +1,51 @@
+package pack;
```
Comment (adriaanm) [inline test/files/run/t7214.scala]: getting rid of the package will probably fix the test failure

**Comment by adriaanm** [comment]
LGTM -- will resubmit with my proposed fix if it does the trick

**Comment by retronym** [comment]
Actually that reminds me, I hit a second bug if things were in the empty package, but I forgot about it when submitting this PR. And I'm not sure how the test passed for me locally. But leave it with me and I'll check on both accounts.

**Comment by retronym** [comment]
Test case fixed. I think I merged a `pos` and `run` test at the last minute before submitting the PR, sorry about the noise.

I can't reproduce the empty package problem. Will keep looking, but it seemed like an unrelated problem, so re-review at will.

**Comment by retronym** [comment]
Okay, the "empty package problem" seems to just be a different manifestation of this bug; both are fixed.

```
ticket/7214 ~/code/scala2 head -n1 test/files/run/t7214.scala // pattern matcher crashes here trying to synthesize an uneeded outer test.
ticket/7214 ~/code/scala2 qbin/scalac test/files/run/t7214.scalaticket/7214 ~/code/scala2 scalac210 test/files/run/t7214.scalatest/files/run/t7214.scala:16: error: error during expansion of this match (this is a scalac bug).
The underlying error was: type mismatch;
 found   : <notype>
 required: AnyRef
Note that <none> extends Any, not AnyRef.
Such types can participate in value classes, but instances
cannot appear in singleton types or in reference comparisons.
  (null: Any) match {
              ^
one error found
ticket/7214 ~/code/scala2 head -n1 test/files/run/t7214.scala 
package pack
ticket/7214 ~/code/scala2 qbin/scalac test/files/run/t7214.scala
ticket/7214 ~/code/scala2 scalac210 test/files/run/t7214.scala
error: no-symbol does not have an owner

     while compiling: test/files/run/t7214.scala
        during phase: patmat
     library version: version 2.10.0-RC1
    compiler version: version 2.10.0-RC1
  reconstructed args: 
```

**Comment by adriaanm** [comment]
Excellent. Merging.

**Pull request merged**